### ISSUE TITLE
USB Kmod packages for this release

### ISSUE THREAD

**Issue by nullPointerMV** [gwlim/Fast-Path-LEDE-OpenWRT#3]
Hi, 

Could you either include support for, or provide me compiled packages, for the kernel versions associated with the builds for WDR4300 and Archer C7v2 for usb-storage and its related requisites for extroot?
As well, (if it's possible), the f2fs support? Not sure if the mips24k or mips74kc architectures support it (there's a few differing opinions on this, from what I've seen, and I just get a "bus error" with a different/older firmware that had USB-storage support, but did not have the same issues formatting the usb-key as ext4

The kernel version in these builds is newer than that in the repo, so even if I force --nodeps, and either link/copy the downloaded mods to the current-kernel, they won't insmod b/c of unresolved/unknown symbols, and if I browse LEDE's dev/nightly package list, it has a newer kernel version than the Sept. builds, so same issue

I'd really like to setup/install extroot on my devices, to give a bit more room than the stock flash size

**Comment by asergiu** [comment]
yep same here. Is there any other way modules could be added without rebuilding the entire image ?
It seems the septmeber build is based on a newer kernel than that of 17.0.1.2 - so even if manually installing kernel modules and solving depends manually (other modules) they will not load. 

scsi_mod.ko (kmod-scsi-core) is missing a symbol: _scsi_mod: Unknown symbol warn_slowpath_null_ right now

**Comment by gwlim** [comment]
Yes I know, but I am really busy this month so I am not able to commit. I can only resume end of next month.
Sorry about that.

**Comment by gwlim** [comment]
Kernel Config issues should be fixed hopefully when I tried to install kmod-scsi-core on the latest image it seems okay
Network Kmods are toasted because of the additional fast path kernel symbols

**Comment by jgvega** [comment]
Hi.
I`m interested to image for TP-link wrd3600 for to use ext-root.


The necessary modules for the EXT-root are:
- kmod-usb-storage (Kernel modules -> USB support -> kmod-usb-storage)
- kmod-fs-ext4 (Kernel modules -> Filesystems -> kmod-fs-ext4)
- kmod-fs-f2fs (recommended for usb flash. Kernel modules -> Filesystems -> kmod-fs-f2fs )
- block-mount (Base system->block-mount)
- kmod-scsi-core (Kernel modules -> Block device-> k-mod-scsi-core)

Optional for compatibility Microsoft format are:
- kmod-fs-msdos (Kernel modules -> Filesystems -> kmod-fs-msdos )
- kmod-fs-ntfs (Kernel modules -> Filesystems -> kmod-fs-ntfs )
- kmod-fs-vfat (Kernel modules -> Filesystems -> kmod-fs-vfat)

I modify your September script and it works well (only path 066-... not work). Checkout for kernel 4.4.87, actually is kernel 4.4.102 and not work.

`git clone https://github.com/lede-project/source.git lede`

`git clone -b lede-17.01 https://github.com/gwlim/mips74k-lede-patch.git temp;mv temp/patch/066-wireguard-update.patch temp/broken; mv temp/* lede/; rm -rf temp`

`cd lede`

`git checkout ab305e147ed7f950e8b34d0de23797c8dbb6a27f`

`./patch_LEDE.sh`

 make menuconfig

Regards.

**Comment by pekseg** [comment]
jgvega,

I followed the build procedure, (build was successful) and flashed my 1043nd v1 
and it won't boot.

Now I need to make a serial connection....so be careful, something is odd.

**Comment by jgvega** [comment]
Hi.
Your model router need mips24k-lede-patch nota mips74k.
My script is only for mips74k. You need to modify my script for your model router.
You bricked your router because you used image for other model router.

**Comment by asergiu** [comment]
Hello,    Is anyone still working from time to time on this ? How about its status integration on openwrt - I have seen it has some implications on some Qos algos which do not work anymore - or it is not the case anymore ?
I do have a router with the gwlim build on it - but cannot add kernel modules as packages as they do not match - the build being too old.
Adrian
    On Tuesday, 9 January 2018, 16:24:16 EET, jgvega <notifications@github.com> wrote:  
 
 
Hi.
Your model router need mips24k-lede-patch nota mips74k.
My script is only for mips74k. You need to modify my script for your model router.
You bricked your router because you used image for other model router.

**Comment by jgvega** [comment]
Hi.
Each compilation of the kernel has a hash code.
When you make a customized image, the kernel modules can only be installed when the image is built in menuconfig.

Packages that are not kernel modules can be installed with opkg, but the use of insmod is not possible.

regards

**Comment by asergiu** [comment]
Yes I know about the kernel module problem. I was wandering if any progress has been made to integrate the patch in lede or not ?
Otherwise we need to build images ourselves whenever there are security patches or enhancements to original lede source tree.
Adrian
    On Wednesday, 10 January 2018, 13:58:48 EET, jgvega <notifications@github.com> wrote:  
 
 
Hi.
Each compilation of the kernel has a hash code.
The kernel modules can only be installed when the image is built in menuconfig.

Packages that are not kernel modules can be installed with opkg, but the use of insmod is not possible.

regards

**Comment by juppin** [comment]
I had prebuild some images with latest 17.01 which support extroot for ext4 and f2fs.
Kernel version is 4.4.120.

The images are about 7MB but i had also added some extra packages such as full dnsmasq, nano...

Look at file included_packages.txt in my fork.

[My fork](https://github.com/juppin/Fast-Path-LEDE-OpenWRT/tree/master/Fast-Path/Mar-2018)

If you need a build for a other router, feel free to write a response.
Or feel free to use fixed patches by me [mips74k-patch](https://github.com/juppin/mips74k-ar71xx-lede-patch).

Greets

**Comment by jgvega** [comment]
Hi.
I compiled for my TP-link wrd3600 for to use ext-root. 
When I get home I compare my version of packages included with yours.

Regards.

**Comment by elgab** [comment]
With "Sep-2018" installing kmods still produced bootloop. Can USB / ext-root modules be included as optional images for WDR's?

**Comment by gwlim** [comment]
That is right. I have upgraded the compiler and kernel and optimizations too far from the original LEDE so ABI changes cause the kmod no longer to work. I can add USB Modules but personally I find it quite pointless, filesharing is not suppose to be the duty of a networking device. Even if you can mount a drive off the router, it is not going to be fast, during file transfer the RAM usage will kill the router defeating the original purpose of a high performance network device, what do you think?
I have really tune performance to be really above original openwrt you can probably feel it.

**Comment by asergiu** [comment]
About optimizations @gwlim ? You also put alot of CPU specific optimizations in the compiler options (arguments). Do they really impact that much the end result ? I am asking this as on your builds tcpdump crashes every few seconds with **Bus error**:
_root@Tvarita:~# tcpdump -i br-lan
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on br-lan, link-type EN10MB (Ethernet), capture size 262144 bytes
07:58:05.125800 ARP, Request who-has Tvarita.lan tell iPhone.lan, length 28
Bus error_

Only cause of those crashes I could find is because of too agressive compiler optimizations...

Some kernel modules bootloop the device (wdr4300 v1 in my case) - kmod-nf-nathelper-extra package for pptp and others, etc. This worked in a previous image build.

I know is impossible to test on too many devices but having  a working image that is closer to stock lede/openwrt and slowly improve it as the user base grows and there are many testers could be a better alternative  than going directly and trying to get every single feature of a CPU architecture

Not criticising here ..... anyway - great work otherwise

**Comment by gwlim** [comment]
The bus error issue is not related with the kmod issue. 
Bus error issue has been fixed and tested with the revise image.
I have tried to fix the kmod issue but the toolchain/kernel update difference is too big to resolve
